storage/spanstore/downsampling_writer.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanstore
+
+import (
+	"hash"
+	"hash/fnv"
+	"math"
+	"sync"
+
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/jaegertracing/jaeger/model"
+)
+
+const (
+	traceIDByteSize = 16
+)
+
+// hasher includes data we want to put in sync.Pool
+type hasher struct {
+	hash   hash.Hash64
+	buffer []byte
+}
+
+// downsamplingWriterMetrics keeping track of total number of dropped spans and accepted spans
+type downsamplingWriterMetrics struct {
+	SpansDropped  metrics.Counter `metric:"spans_dropped"`
+	SpansAccepted metrics.Counter `metric:"spans_accepted"`
+}
+
+// DownsamplingWriter is a span Writer that drops spans with a predefined downsamplingRatio.
+type DownsamplingWriter struct {
+	spanWriter   Writer
+	threshold    uint64
+	lengthOfSalt int
+	hasherPool   *sync.Pool
+	metrics      downsamplingWriterMetrics
+}
+
+// DownsamplingOptions contains the options for constructing a DownsamplingWriter.
+type DownsamplingOptions struct {
+	Ratio          float64
+	HashSalt       string
+	MetricsFactory metrics.Factory
+}
+
+// NewDownsamplingWriter creates a DownsamplingWriter.
+func NewDownsamplingWriter(spanWriter Writer, downsamplingOptions DownsamplingOptions) *DownsamplingWriter {
+	threshold := uint64(downsamplingOptions.Ratio * float64(math.MaxUint64))
+	writeMetrics := &downsamplingWriterMetrics{}
+	metrics.Init(writeMetrics, downsamplingOptions.MetricsFactory, nil)
+	hashSaltBytes := []byte(downsamplingOptions.HashSalt)
```

Comment:
I would prefer to define a private constant for defaultSalt & initialize it to some random string, and use that when `downsamplingOptions.HashSalt  == ""`. This would avoid the possibility that some other trace id hashing solution elsewhere would also run with empty string as a salt.

Reply:
If you mean same non-empty salt passed from CreateSpanWriter to DownsamplingWriter, I would pass it in `factory.go`. 

Else we would end up with different salt for different ingesters which won't guarantee that we always downsampling/ not downsampling spans from same trace (if they land on different ingesters)?

Reply:
why would it be different if it's a hardcoded const value, only used when no salt is provided externally?

Reply:
i think i misunderstood your meaning. Updated the code and passed a hardcoded const value from `factory.go` for `downsamplingWriter`.

Reply:
why not just pass the const from factory.go? This way if downsamplingWriter is used elsewhere we could just pass different const string.

Reply:
because right now your default in the factory is also `""`. What I am suggesting does not prevent using the writer with a different salt:
```
const defaultHashSalt  = "Bender is great!"

func NewDownsamplingWriter(spanWriter Writer, downsamplingOptions DownsamplingOptions) *DownsamplingWriter {
    salt := downsamplingOptions. HashSalt
    if salt == "" {
        salt = defaultHashSalt
    }
    // the rest of constructor
}
```